apis/python/src/tiledbsc/annotation_dataframe.py

```diff
@@ -0,0 +1,118 @@
+        """
+        Populates the obs/ or var/ subgroup for a SOMA object.
+
+        :param dataframe: anndata.obs, anndata.var, anndata.raw.var.
+        :param extent: TileDB extent parameter for the array schema.
+        """
+
+        offsets_filters = tiledb.FilterList(
+            [tiledb.PositiveDeltaFilter(), tiledb.ZstdFilter(level=-1)]
+        )
+        dim_filters = tiledb.FilterList([tiledb.ZstdFilter(level=-1)])
+        attr_filters = tiledb.FilterList([tiledb.ZstdFilter(level=-1)])
+
+        if self.verbose:
+            s = util.get_start_stamp()
+            print(f"{self.indent}START  WRITING {self.uri}")
+
+        # Make the row-names column (barcodes for obs, gene names for var) explicitly named.
+        # Otherwise it'll be called '__tiledb_rows'.
+        #
+        # Before:
+        #
+        #   >>> anndata.obs
+        #                  orig.ident nCount_RNA nFeature_RNA ...
+        #   ATGCCAGAACGACT 0          70.0       47           ...
+        #   CATGGCCTGTGCAT 0          85.0       52           ...
+        #   ...            ...        ...        ...          ...
+        #   GGAACACTTCAGAC 0          150.0      30           ...
+        #   CTTGATTGATCTTC 0          233.0      76           ...
+        #
+        # After:
+        #
+        #   >>> anndata.obs.rename_axis('obs_id')
+        #                  orig.ident nCount_RNA nFeature_RNA ...
+        #   obs_id
+        #   ATGCCAGAACGACT 0          70.0       47           ...
+        #   CATGGCCTGTGCAT 0          85.0       52           ...
+        #   ...            ...        ...        ...          ...
+        #   GGAACACTTCAGAC 0          150.0      30           ...
+        #   CTTGATTGATCTTC 0          233.0      76           ...
+        dataframe = dataframe.rename_axis(self.name+'_id')
+
+        mode = 'ingest'
+        if self.exists():
+            mode = 'append'
+            if self.verbose:
+                print(f"{self.indent}Re-using existing array {self.uri}")
+
+        tiledb.from_pandas(
+            uri=self.uri,
+            dataframe=dataframe,
+            name=self.name,
+            sparse=True,
+            allows_duplicates=False,
+            offsets_filters=offsets_filters,
+            attr_filters=attr_filters,
+            dim_filters=dim_filters,
+            capacity=100000,
+            tile=extent,
+            ctx=self.ctx,
+            mode=mode,
+        )
+
+        if self.verbose:
+            print(util.format_elapsed(s, f"{self.indent}FINISH WRITING {self.uri}"))
+
+
+    # ----------------------------------------------------------------
+    def to_dataframe(self) -> pandas.DataFrame:
+        """
+        Reads the TileDB obs or var array and returns a type of pandas dataframe
+        and dimension values.
+        """
+
+        if self.verbose:
+            s = util.get_start_stamp()
+            print(f"{self.indent}START  read {self.uri}")
+
```

Comment:
with tiledb.open(self.uri) as A:
    retval = A.df[:]

The `retval` DataFrame should already have index `obs_id`/`var_id`, so no need to set and name it separately.

Reply:
@aaronwolen nice, thanks! :)

Reply:
... alas,

```
KeyError: b'AAATTCGAATCACG'
```

on outgest -- issue is with the 'before' code the df is like

```
                   orig.ident  nCount_RNA  nFeature_RNA  RNA_snn_res.0.8  letter.idents groups  RNA_snn_res.1
obs_id
b'AAATTCGAATCACG'           0       327.0            62                1              1     g2              1
b'AAGCAAGAGCTTAG'           0       126.0            48                0              0     g1              0
b'AAGCGACTTTGACG'           0       443.0            77                1              1     g1              1
b'AATGCGTGGACGGA'           0       389.0            73                1              1     g1              1
b'AATGTTGACAGTCA'           0       100.0            41                0              0     g1              0
...                       ...         ...           ...              ...            ...    ...            ...
b'TTACGTACGTTCAG'           0       228.0            39                0              0     g1              0
b'TTGAGGACTACGCA'           0       787.0            88                0              0     g1              2
b'TTGCATTGAGCTAC'           0       104.0            40                0              0     g2              2
b'TTGGTACTGAATCC'           0       135.0            45                0              0     g1              2
b'TTTAGCTGTACTCT'           0       462.0            86                1              1     g1              1
```

and after

```
                orig.ident  nCount_RNA  nFeature_RNA  RNA_snn_res.0.8  letter.idents groups  RNA_snn_res.1
obs_id
AAATTCGAATCACG           0       327.0            62                1              1     g2              1
AAGCAAGAGCTTAG           0       126.0            48                0              0     g1              0
AAGCGACTTTGACG           0       443.0            77                1              1     g1              1
AATGCGTGGACGGA           0       389.0            73                1              1     g1              1
AATGTTGACAGTCA           0       100.0            41                0              0     g1              0
...                    ...         ...           ...              ...            ...    ...            ...
TTACGTACGTTCAG           0       228.0            39                0              0     g1              0
TTGAGGACTACGCA           0       787.0            88                0              0     g1              2
TTGCATTGAGCTAC           0       104.0            40                0              0     g2              2
TTGGTACTGAATCC           0       135.0            45                0              0     g1              2
TTTAGCTGTACTCT           0       462.0            86                1              1     g1              1
```

namely a bytes-vs-strings difference. I'll add a clarifying comment.

Thank you!! :D